### PR TITLE
Simplify Stage0 page table assembly code

### PR DESCRIPTION
**Pull request by conradgrobler**
This PR simplifies the Stage 0 assembly code related to page tables:

- Move the page tables to BSS
- Move the code to zeo-out BSS earlier so we don't have to separately zero out the page tables
- Only map 2MiB huge pages in assembly and create the 4KiB pages in Rust code
- Only update the actual page table entries we created with the encrypted bit rather than looping through all the page table entries

### CODE REVIEW

File: stage0/src/asm/boot.s [hunk 1]
```diff
@@ -167,33 +142,24 @@ _protected_mode_start:
...
-    or %edi, 4(%ebp,%esi,8)   # PT_0[ESI] |= EDI
-    inc %esi                  # ESI = ESI + 1
-    cmp $512, %esi            # have we changed all 512 entries?
-    jne 1b                    # ESI < 512, go back to square 1
-    pop %esi                  # keep track of the encrypted bit for main
+    shl %cl, %edi             # construct the encrypted bit mask, store it in EDI
+
+    # We set the encrypted bit for each of the page table entries that we previously created.
+    # The encrypted bit is in the second half of each 8-byte entry, so we add an extra offset of 4 bytes.
+    mov ${pml4}, %eax
```
Comment (andrisaar) [inline stage0/src/asm/boot.s:153]: I think you can collapse these down to just one `mov`, just like with the first half.

File: stage0/src/asm/boot.s [hunk 2]
```diff
@@ -68,49 +68,39 @@ _protected_mode_start:
...
-
-    mov ${pd_3}, %edi
-    rep stosb
-
-    mov ${pt_0}, %edi
-    rep stosb
+    # Clear BSS: base address goes to RDI, value (0) goes to AX,
+    # count/4 (we are updating 4 bytes at a time) goes into CX.
+    mov $bss_start, %edi
+    mov $bss_size, %ecx
```
Comment (andrisaar) [inline stage0/src/asm/boot.s:80]: could there be a risk of leaving some garbage in the memory if BSS is not exactly divisible by 4?

File: stage0/src/asm/boot.s [hunk 3]
```diff
@@ -68,49 +68,39 @@ _protected_mode_start:
     # Switch to a flat 32-bit data segment, giving us access to all 4G of memory.
     mov $ds, %eax
     mov %eax, %ds
+    mov %eax, %es
```
Comment (andrisaar) [inline stage0/src/asm/boot.s:71]: We shouldn't be using ES anywhere -- why is this needed?

File: stage0/src/asm/boot.s [hunk 4]
```diff
@@ -167,33 +142,24 @@ _protected_mode_start:
     xor %ecx, %ecx            # ECX = 0 - we're not interested in a subpage
     cpuid                     # EAX, EBX, ECX, EDX = CPUID(EAX, ECX)
     and $0b111111, %ebx       # zero out all but EBX[5:0], which the C-bit location
-    push %ebx                 # save the full C-bit location value for later
+    mov %ebx, %esi            # save the full C-bit location value for later to pass into main
```
Comment (andrisaar) [inline stage0/src/asm/boot.s:145]: We no longer need  EDI for any particular purpose (like `rep stosb`), so instead of putting the C-bit location in `ESI` here and moving it into `EDI` later, we could just put it in `EDI` directly here and use `ESI` for the bitmask below.

**Review by andrisaar — APPROVED** [review]

File: stage0/src/asm/boot.s (hunk 4 above)
Comment (conradgrobler) [reply stage0/src/asm/boot.s:145]: Good point. I will change that.

**Review by conradgrobler — COMMENTED** [review]
Thanks
